File: src/main/java/com/gmail/borlandlp/minigamesdtools/nmsentities/entity/SkyDragon.java
package com.gmail.borlandlp.minigamesdtools.nmsentities.entity;

import com.gmail.borlandlp.minigamesdtools.nmsentities.classes.PathProvider;
import com.gmail.borlandlp.minigamesdtools.nmsentities.classes.SkyDragonControllerCharge;
import net.minecraft.server.v1_12_R1.*;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.craftbukkit.v1_12_R1.CraftServer;
import org.bukkit.craftbukkit.v1_12_R1.CraftWorld;
import org.bukkit.craftbukkit.v1_12_R1.entity.CraftEnderDragon;
import org.bukkit.entity.EnderDragon;
import org.bukkit.event.entity.CreatureSpawnEvent;

import java.lang.reflect.Field;

public class SkyDragon extends EntityEnderDragon implements IMoveControllableEntity, IControllable, ISkybattleEntity {
    private CraftEnderDragon bukkitEntity;
    private DragonControllerManager dragonControllerManager;

    public SkyDragon(org.bukkit.World world) {
        super((World) ((CraftWorld) world).getHandle());
        this.children = new EntityComplexPart[]{this.bw, this.bx, this.by, this.bz, this.bA, this.bB, this.bC, this.bD};
        this.setHealth(this.getMaxHealth());
        this.setSize(16.0F, 8.0F);
        this.noclip = true;
        this.fireProof = true;
        this.setNoAI(false);


        try {
            Field fbM = EntityEnderDragon.class.getDeclaredField("bM");
            fbM.setAccessible(true);
            fbM.setInt(this, 100);

            Field fbK = EntityEnderDragon.class.getDeclaredField("bK");
            fbK.setAccessible(true);
            if (!((CraftWorld) world).getHandle().isClientSide && ((CraftWorld) world).getHandle().worldProvider instanceof WorldProviderTheEnd) {
                fbK.set(this, ((WorldProviderTheEnd)((CraftWorld) world).getHandle().worldProvider).t());
            } else {
                fbK.set(this, null);
            }

            this.dragonControllerManager = new DragonControllerManager(this);
            Field fbL = EntityEnderDragon.class.getDeclaredField("bL");
            fbL.setAccessible(true);
            fbL.set(this, this.getDragonControllerManager());
        } catch (NoSuchFieldException e1) {
            e1.printStackTrace();
        } catch (IllegalAccessException e1) {
            e1.printStackTrace();
        }


        this.ah = true;
        this.bukkitEntity = new CraftEnderDragon((CraftServer) Bukkit.getServer(), this);
    }

    @Override
    public DragonControllerManager getDragonControllerManager() {
        return dragonControllerManager;
    }

    @Override
    public void setPath(PathProvider path) {
        DragonControllerPhase dragoncontrollerphase = CraftEnderDragon.getMinecraftPhase(CraftEnderDragon.getBukkitPhase(DragonControllerPhase.i));
        //IDragonController iDragonController = dragoncontrollerphase.a(this);

        Field curDragContr = null;
        try {
            curDragContr = this.getDragonControllerManager().getClass().getDeclaredField("currentDragonController");
        } catch (NoSuchFieldException e1) {
            e1.printStackTrace();
        }
        curDragContr.setAccessible(true);
        try {
            curDragContr.set(this.getDragonControllerManager(), (SkyDragonControllerCharge) path);
        } catch (IllegalAccessException e1) {
            e1.printStackTrace();
        }
        this.getDataWatcher().set(EntityEnderDragon.PHASE, dragoncontrollerphase.b());
    }

    @Override
    public PathProvider vec2Path(Vec3D target) {
        SkyDragonControllerCharge iDragonController = new SkyDragonControllerCharge(this);
        iDragonController.a(target);

        return iDragonController;
    }

    @Override
    public boolean isFollowsThisPath(PathProvider pathProvider) {
        return this.getDragonControllerManager().a() == pathProvider;
    }

    @Override
    public CraftEnderDragon getBukkitEntity()
    {
        return bukkitEntity;
    }

    protected void initAttributes()
    {
        super.initAttributes();
    }

    public EnderDragon spawn(Location loc) {
        this.setLocation(loc.getX(), loc.getY(), loc.getZ(), loc.getYaw(), loc.getPitch());
        this.world.addEntity(this, CreatureSpawnEvent.SpawnReason.CUSTOM);

        return getBukkitEntity();
    }

    @Override
    public void teleport(PathProvider pathProvider) {
        org.bukkit.World world = this.getBukkitEntity().getLocation().getWorld();
        Location location = new Location(world, pathProvider.getPath().x, pathProvider.getPath().y, pathProvider.getPath().z);
        this.getBukkitEntity().teleport(location);
    }

    //onLivingUpdate();
    /*public void n() {
        // dragon living
    }*/
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/nmsentities/entity/SkyDragon.kt
package com.gmail.borlandlp.minigamesdtools.nmsentities.entity

import com.gmail.borlandlp.minigamesdtools.nmsentities.classes.PathProvider
import com.gmail.borlandlp.minigamesdtools.nmsentities.classes.SkyDragonControllerCharge
import net.minecraft.server.v1_12_R1.*
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.World
import org.bukkit.craftbukkit.v1_12_R1.CraftServer
import org.bukkit.craftbukkit.v1_12_R1.CraftWorld
import org.bukkit.craftbukkit.v1_12_R1.entity.CraftEnderDragon
import org.bukkit.entity.EnderDragon
import org.bukkit.event.entity.CreatureSpawnEvent
import java.lang.reflect.Field

class SkyDragon(world: World) :
    EntityEnderDragon((world as CraftWorld).handle as net.minecraft.server.v1_12_R1.World),
    IMoveControllableEntity, IControllable, ISkybattleEntity {

    private val wrappedBukkitEntity: CraftEnderDragon
    private var dragonControllerManager: DragonControllerManager? = null
    override fun getDragonControllerManager(): DragonControllerManager {
        return dragonControllerManager!!
    }

    override fun setPath(path: PathProvider) {
        val dragoncontrollerphase =
            CraftEnderDragon.getMinecraftPhase(CraftEnderDragon.getBukkitPhase(DragonControllerPhase.i))
        //IDragonController iDragonController = dragoncontrollerphase.a(this);
        var curDragContr: Field? = null
        try {
            curDragContr = getDragonControllerManager().javaClass.getDeclaredField("currentDragonController")
        } catch (e1: NoSuchFieldException) {
            e1.printStackTrace()
        }
        curDragContr!!.isAccessible = true
        try {
            curDragContr[getDragonControllerManager()] = path as SkyDragonControllerCharge
        } catch (e1: IllegalAccessException) {
            e1.printStackTrace()
        }
        this.dataWatcher.set(PHASE, dragoncontrollerphase.b())
    }

    override fun vec2Path(vec3D: Vec3D): PathProvider {
        val iDragonController = SkyDragonControllerCharge(this)
        iDragonController.a(vec3D)
        return iDragonController
    }

    override fun isFollowsThisPath(pathProvider: PathProvider): Boolean {
        return getDragonControllerManager().a() === pathProvider
    }

    override fun getBukkitEntity(): CraftEnderDragon {
        return wrappedBukkitEntity
    }

    override fun initAttributes() {
        super.initAttributes()
    }

    override fun spawn(loc: Location): EnderDragon {
        setLocation(loc.x, loc.y, loc.z, loc.yaw, loc.pitch)
        world.addEntity(this, CreatureSpawnEvent.SpawnReason.CUSTOM)
        return getBukkitEntity()
    }

    override fun teleport(pathProvider: PathProvider) {
        val world = getBukkitEntity().location.world
        val location =
            Location(world, pathProvider.path!!.x, pathProvider.path!!.y, pathProvider.path!!.z)
        getBukkitEntity().teleport(location)
    } //onLivingUpdate();

    /*public void n() {
        // dragon living
    }*/
    init {
        children =
            arrayOf(bw, bx, by, bz, bA, bB, bC, bD)
        this.health = this.maxHealth
        setSize(16.0f, 8.0f)
        noclip = true
        fireProof = true
        this.isNoAI = false
        try {
            val fbM = EntityEnderDragon::class.java.getDeclaredField("bM")
            fbM.isAccessible = true
            fbM.setInt(this, 100)
            val fbK = EntityEnderDragon::class.java.getDeclaredField("bK")
            fbK.isAccessible = true
            if (!(world as CraftWorld).handle.isClientSide && world.handle.worldProvider is WorldProviderTheEnd) {
                fbK[this] = (world.handle.worldProvider as WorldProviderTheEnd).t()
            } else {
                fbK[this] = null
            }
            dragonControllerManager = DragonControllerManager(this)
            val fbL = EntityEnderDragon::class.java.getDeclaredField("bL")
            fbL.isAccessible = true
            fbL[this] = getDragonControllerManager()
        } catch (e1: NoSuchFieldException) {
            e1.printStackTrace()
        } catch (e1: IllegalAccessException) {
            e1.printStackTrace()
        }
        ah = true
        this.wrappedBukkitEntity = CraftEnderDragon(Bukkit.getServer() as CraftServer, this)
    }
}